cpp/src/lists/apply_boolean_mask.cu

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/copy.hpp>
+#include <cudf/detail/fill.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/reduction_functions.hpp>
+#include <cudf/detail/replace.hpp>
+#include <cudf/detail/stream_compaction.hpp>
+#include <cudf/lists/detail/stream_compaction.hpp>
+#include <cudf/lists/stream_compaction.hpp>
+#include <cudf/utilities/bit.hpp>
+
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/reduce.h>
+
+namespace cudf::lists {
+namespace detail {
+
+/**
+ * @copydoc cudf::lists::detail::apply_boolean_mask
+ */
+std::unique_ptr<column> apply_boolean_mask(lists_column_view const& input,
+                                           lists_column_view const& boolean_mask,
+                                           rmm::cuda_stream_view stream,
+                                           rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(boolean_mask.child().type().id() == type_id::BOOL8, "Mask must be of type BOOL8.");
+  CUDF_EXPECTS(input.size() == boolean_mask.size(),
+               "Boolean masks column must have same number of rows as input.");
+  auto const num_rows = input.size();
+
+  if (num_rows == 0) { return cudf::empty_like(input.parent()); }
+
+  auto constexpr offset_data_type = data_type{type_id::INT32};
+
+  auto filtered_child = [&] {
+    auto filtered =
+      cudf::detail::apply_boolean_mask(cudf::table_view{{input.get_sliced_child(stream)}},
+                                       boolean_mask.get_sliced_child(stream),
+                                       stream,
+                                       mr)
+        ->release();
+    return std::move(filtered.front());
+  };
+
+  auto output_offsets = [&] {
+    auto boolean_mask_sliced_offsets =
+      cudf::detail::slice(
+        boolean_mask.offsets(), {boolean_mask.offset(), boolean_mask.size() + 1}, stream)
+        .front();
+    auto const sizes       = cudf::reduction::segmented_sum(boolean_mask.get_sliced_child(stream),
```

Comment:
If not for the lambdas you would be able to reuse the output of `boolean_mask.get_sliced_child(stream)`. As it stands now, you could still avoid doing this twice by pulling it out of the lambdas and storing it in the function scope then referencing it in the lambdas. I believe each call requires a `cudaMemcpyAsync` under the hood, which would be nice to avoid.

Reply:
I suppose one reason to avoid doing this operation through `segmented_sum` would be to avoid having to materialize the sliced child offsets and instead be able to use the iterator that accounts for the column-level offset in the offsets child.